File: python/exp2_pipeline.py
import numpy as np

from common import MiB, append_csv, benchmark, make_bytes

HEADER = 64
sizes = [1, 2, 4, 8, 16, 32, 64]


def copy_pipeline(packet, body_size):
    payload = packet[HEADER:HEADER + body_size]
    middle = payload[body_size // 4:]
    copied = bytearray(middle)
    arr = np.frombuffer(copied, dtype=np.uint8).copy()
    return int(arr.sum())


def view_pipeline(packet, body_size):
    view = memoryview(packet)
    payload = view[HEADER:HEADER + body_size]
    middle = payload[body_size // 4:]
    arr = np.frombuffer(middle, dtype=np.uint8)
    return int(arr.sum())


for size_mib in sizes:
    body_size = size_mib * MiB
    packet = make_bytes(body_size + HEADER)

    for name, fn in [
        ("copy_pipeline", lambda: copy_pipeline(packet, body_size)),
        ("view_pipeline", lambda: view_pipeline(packet, body_size)),
    ]:
        ms = benchmark(fn)

        append_csv(
            "exp2.csv",
            {
                "language": "python",
                "experiment": "pipeline",
                "operation": name,
                "size_mib": size_mib,
                "median_ms": ms,
                "throughput_mib_s": size_mib / (ms / 1e3),
            },
        )

File: rust/src/bin/exp2_pipeline.rs
use zero_copy_lab::common::*;

fn copy_pipeline(
    data: &[u8]
) -> u64 {
    let payload = data.to_vec();

    let middle =
        payload[
            payload.len() / 8..
            payload.len() * 7 / 8
        ].to_vec();

    middle.iter().map(|v| *v as u64).sum()
}

fn view_pipeline(
    data: &[u8]
) -> u64 {
    let middle =&data[data.len() / 8.. data.len() * 7 / 8];
    middle.iter().map(|v| *v as u64).sum()
}

fn main() {
    let sizes = vec![1, 2, 4, 8, 16, 32, 64];

    for size_mib in sizes {
        let size = size_mib * MIB;

        println!("Running Rust exp2: {} MiB", size_mib);

        let data = vec![1u8; size];

        let copy_ms =
            benchmark(
                || {
                    std::hint::black_box(
                        copy_pipeline(&data)
                    );
                },
                10,
            );

        let view_ms =
            benchmark(
                || {
                    std::hint::black_box(
                        view_pipeline(&data)
                    );
                },
                10,
            );

        append_csv(
            "../results/exp2.csv",
            &BenchmarkRow {
                language:
                    "rust".into(),
                experiment:
                    "pipeline".into(),
                operation:
                    "copy_pipeline"
                        .into(),
                size_mib,
                median_ms:
                    copy_ms,
                throughput_mib_s:
                    size_mib as f64
                        / ((copy_ms + 1e-9)
                            / 1000.0),
            },
        );

        append_csv(
            "../results/exp2.csv",
            &BenchmarkRow {
                language:
                    "rust".into(),
                experiment:
                    "pipeline".into(),
                operation:
                    "view_pipeline"
                        .into(),
                size_mib,
                median_ms:
                    view_ms,
                throughput_mib_s:
                    size_mib as f64
                        / ((view_ms + 1e-9)
                            / 1000.0),
            },
        );
    }
}